### PR TITLE
[Feature] Add MLAProcess for DeepSeek MLA on NPU

### PR DESCRIPTION
**Pull request by iforgetmyname**
## Motivation

This pr introduced a new fused op that warps whole `forward_absorb_prepare` function in mla. Codes for this op is open-sourced [here](https://github.com/sgl-project/sgl-kernel-npu/pull/51), docs on the way.

## Modifications

- Add a new fused op

## Accuracy Tests

image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Review by merrymercy — APPROVED** [review]

File: python/sglang/srt/layers/attention/npu_ops/mla_preprocess.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+            fused_qkv_a_proj_with_mqa_weight_new.t().contiguous()
+        )
+        fused_qkv_a_proj_with_mqa_weight_nz = (
+            transdata(fused_qkv_a_proj_with_mqa_weight, block_size=(16, 32))
+            .unsqueeze(0)
+            .contiguous()
+        )
+        self.qkv_a_proj_weight_nz = torch_npu.npu_format_cast(
+            fused_qkv_a_proj_with_mqa_weight_nz, 29
+        )
```
[the remark this answers is not in the document]
Comment (iforgetmyname) [reply python/sglang/srt/layers/attention/npu_ops/mla_preprocess.py:127]: done

File: python/sglang/srt/layers/attention/npu_ops/mla_preprocess.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+        q_b_proj_quant_bias = q_b_proj_quant_bias.reshape(
+            self.num_local_heads, self.qk_nope_head_dim + self.qk_rope_head_dim, -1
+        )
+        q_b_proj_quant_bias = trans_rope_weight(
+            q_b_proj_quant_bias, self.qk_rope_head_dim
+        )
+        self.q_b_proj_quant_bias = q_b_proj_quant_bias.reshape(
+            self.num_local_heads * (self.qk_nope_head_dim + self.qk_rope_head_dim)
+        )
+
```
[the remark this answers is not in the document]
Comment (iforgetmyname) [reply python/sglang/srt/layers/attention/npu_ops/mla_preprocess.py:226]: n/a

File: python/sglang/srt/layers/attention/npu_ops/mla_preprocess.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+def is_mla_preprocess_enabled() -> bool:
+    return _is_npu and _ENABLE_MLA_PREPROCESS_FLAG
+
+
+if is_mla_preprocess_enabled():
+    import sgl_kernel_npu
+    import torch_npu
+
+
+def round_up(val: int, align: int) -> int:
```
Comment (ping1jing2) [inline python/sglang/srt/layers/attention/npu_ops/mla_preprocess.py:23]: please add docstring

File: python/sglang/srt/layers/attention/npu_ops/mla_preprocess.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+    if align == 0:
+        return 0
+    return -(val // -align) * align
+
+
+def transdata(nd_mat, block_size: tuple = (16, 16)):
```
Comment (ping1jing2) [inline python/sglang/srt/layers/attention/npu_ops/mla_preprocess.py:29]: ditto

**Review by ping1jing2 — APPROVED** [review]

File: python/sglang/srt/layers/attention/npu_ops/mla_preprocess.py (hunk 3 above)
Comment (iforgetmyname) [reply python/sglang/srt/layers/attention/npu_ops/mla_preprocess.py:23]: will add soon

File: python/sglang/srt/layers/attention/npu_ops/mla_preprocess.py (hunk 4 above)
Comment (iforgetmyname) [reply python/sglang/srt/layers/attention/npu_ops/mla_preprocess.py:29]: will add soon

**Review by Alcanderian — APPROVED** [review]

File: python/sglang/srt/models/deepseek_v2.py [hunk 5]
```diff
@@ -1263,9 +1273,28 @@ def forward_prepare(
...
-                positions, hidden_states, forward_batch, zero_allocator
-            )
+            if not self.is_mla_preprocess_enabled:
+                inner_state = self.forward_absorb_prepare(
+                    positions, hidden_states, forward_batch, zero_allocator
+                )
+            else:
+                # TODO(iforgetmyname): to be separated as a standalone func
+                if self.mla_preprocess is None:
+                    self.mla_preprocess = NPUFusedMLAPreprocess(
```
Comment (yuan-luo) [inline python/sglang/srt/models/deepseek_v2.py:1283]: NPUFusedMLAPreprocess() should not be in common path, even though the implementation has no NPU specific biz logic.
